feat(web): synchronise metadata and album membership between duplicate images

EinToni opened the pull request:
## Description
This is my first time contributing here and I hope it is ok to open a PR without creating an issue. This PR arises from the feature request #10262 and in part from #10206. In both more information is wished as well as the possibility to synchronize metadata between the duplicated images.

## PR Content
This PR introduces the ability to automatically:
* Synchronize the albums of duplicated images
  * Meaning: The deduplicated images are all put in all albums any duplicate was part of.
* Synchronize the favorites
  * Meaning: If any of the duplicates was marked as favorite, the deduplicated images are also marked as favorites.
* Synchronize the archive
  * Meaning: If any of the duplicates was archived, also archive the deduplicated images.

These three settings can be individually toggled:
![image](https://github.com/user-attachments/assets/9ac2cfb0-a94f-4a93-a5d2-43a95f7071c6)

I also added an indicator for the archive status of any image, same as the favorites icon:
![image](https://github.com/user-attachments/assets/171bf1f9-8378-4638-b355-9568d11a863c)

And lastly I added the possibility to manually select the date/time, description and location to be synched to the deduplicated kept images. I am not that happy with this solution, but these are all fields that could completely differ from each other so no automatic resolution seemed fitting here.
![image](https://github.com/user-attachments/assets/86be2ba0-57a4-4561-b6d1-4bd185777db8)
In this example the remaining two images would be marked as favorites, marked as archived, have the location set to Paris, the description to "thisis some test" and the timestamp to 9. Jan. 2023 Mo, 07:15.

## Open work
The size of the showed duplicates can now vary and therefore not align with each other. This looks very bad, but I didn't find any nice way to fix this (it's my first time using svelte and I'm not doing web development too often).

EinToni committed:
feat(web): Synchronize information from deduplicated images

* Added new settings menu to the the deduplication tab.
* The toggable options in the settings are synchronization of: albums, favorites, ratings, description, visibility and location.
* When synchronizing the albums, the resolved images will be added to all albums of the duplicates.
* When synchronizing the favorite status, the resolved images will be marked as favorite, if at least one selectable image is marked as favorite.
* When synchronizing the ratings, the highest rating from the selectable images will be applied to the resolved image.
* When synchronizing the description, all descriptions from the selectable images will be merged into one description for the resolved image.
* When synchronizing the visibility, the most restrictive visibility setting from the selectable images will be applied to the resolved image.
* When synchronizing the location, if exactly one unique location exists among the selectable images, this location will be applied to the resolved image.
* There is no additional UI for these settings to keep the visual clutter minimal. The settings are applied automatically based on the user's preferences.

bo0tzz commented:
Cool stuff, thank you!

The UI is a little unclear to me at the moment, I'm hoping @alextran1502 has some ideas there. My suggestion would be to separately display another card showing what the deduplicated end result will look like. Just a super quick and dirty browser devtools mockup:
![image](https://github.com/user-attachments/assets/d265972b-c488-4680-8dc0-28deeaefcae9)

EinToni commented:
Thank you for the suggestion! I like it and it looks great in your example. But where should we put the "preview" card when there are more duplicates (like in my initial example)? The best would probably be to put it below the rest, but I don't know how clean the look then would be.

Also after looking at this again after almost a week, the red and green highlights really clutter up the UI. My idea was to highlight the metadata in red that gets "lost", as it will be overwritten by other data. But I would say removing the red highlighting in favor for proper lineup of the fields will help the readability quite a lot.

I will give the lineup of the data another try and remove the red highlights. Afterwards I can put in your preview card to see how it feels in action.

EinToni commented:
I decided against a separate preview asset because I saw an example on my regular immich instance, where some duplicates are looking visually slightly different. There I want to keep e.g. 2 of 5 images. Showing this in a single preview would only confuse as I would have to pick some image to diplay.

I therefore opted to display the preview directly in the selected assets (and only in the selected ones). Previously I highlighted all metadata that is different from the selected one. Now only the selected assets reflect the changes that will be done. In my opinion this has improved the usability a lot.
I also adapted the shape of the highlight to match the rest of the UI.
![image](https://github.com/user-attachments/assets/10a6bc1e-7cfb-46a1-8922-49f5060ab953)

Usability feels better:
https://github.com/user-attachments/assets/c5b02bdb-73e8-4d9f-9e89-7f109552fb07


I would be happy about further feedback :)

bo0tzz commented:
Good stuff!

> There I want to keep e.g. 2 of 5 images. Showing this in a single preview would only confuse as I would have to pick some image to diplay.

What about displaying both?

EinToni commented:
Displaying every selected asset is another "preview" card would clutter up the UI quite quickly. And when selecting/deselecting some duplicates the previews would probably dynamically pop up and vanish, dependent on the amount of selected duplicates. This could lead to a lot of confusion and also distinguishing between the previews to tell which preview relates to which duplicate could then become harder than necessary.
Therefore I would say a direct preview of the changes in the already existing displays is better.

mertalev commented:
Wow, awesome stuff! My thought is that the syncing for albums, favorites, etc. is relatively straightforward while metadata is harder to display and edit in an intuitive way. What do you say about handling the former in this PR and the latter in a separate PR?

EinToni commented:
Thank you!
We can of course split it up, but as the backend behavior is pretty equal, the only real difference is the UI. If we can come to an agreement for the UI, I can also just do it all in this PR.

But if that agreement sounds unrealistic any time soon I'll split it. I would leave that decision up to the immich team.

NicholasFlamy commented:
> I decided against a separate preview asset because I saw an example on my regular immich instance, where some duplicates are looking visually slightly different. There I want to keep e.g. 2 of 5 images. Showing this in a single preview would only confuse as I would have to pick some image to diplay.
> 
> I therefore opted to display the preview directly in the selected assets (and only in the selected ones). Previously I highlighted all metadata that is different from the selected one. Now only the selected assets reflect the changes that will be done. In my opinion this has improved the usability a lot. I also adapted the shape of the highlight to match the rest of the UI. [image]
> 
> Usability feels better: https://github.com/user-attachments/assets/c5b02bdb-73e8-4d9f-9e89-7f109552fb07
> 
> I would be happy about further feedback :)

I will comment, the text is touching the edges of the box in some places, if you could put at least like a 2px margin/padding that would probably do it. (Not too much ofc, we still need content, but just enough that it doesn't touch.)

EinToni commented:
@bo0tzz What will be the further process to continue with this MR?

Gorbush commented:
A with to also sync:
1. Tags of images
2. Star rating of images

bergbria commented on web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:
```diff
@@ -80,11 +85,56 @@
   };
 
-  const handleResolve = async (duplicateId: string, duplicateAssetIds: string[], trashIds: string[]) => {
+  const handleResolve = async (
```
handleDeduplicateAll isn't updated to respect the new settings. Ideally, both bulk and individual resolutions should behave the same.

bergbria replied:
locally hacked up a version that does this. Definitely unpolished and imperfect, but sharing here in case there's anything useful.

I'm not really convinced that 'fetch all the album memberships up front' is the most efficient approach vs requesting per photo. Probably depends on the number of dupes vs overall album size. Might make sense when there are tons of dupes, but otherwise not.

```ts
const handleDeduplicateAll = async () => {
    console.log("dedupe all");
    
    const albumIdToAddedAssets = new Map<string, Set<string>>();
    const assetIdToAlbums = new Map<string, Set<string>>();

    {
      const relevantAssetIds = new Set<string>();
      duplicates.forEach((group) => group.assets.forEach((asset) => relevantAssetIds.add(asset.id)));

      const allAlbums1 = await getAllAlbums({});
      const allAlbums = await Promise.all(allAlbums1.map((album) => getAlbumInfo({
          id: album.id,
          withoutAssets: false
        })));
      allAlbums.forEach((album) => {
        albumIdToAddedAssets.set(album.id, new Set<string>());
        album.assets.forEach((asset) => {
          if (relevantAssetIds.has(asset.id)) {
            if (!assetIdToAlbums.has(asset.id)) {
              assetIdToAlbums.set(asset.id, new Set<string>());
            }

            assetIdToAlbums.get(asset.id)?.add(album.id);
          }

        })
      });
    }

    debugger;

    let newFavorites = new Array();
    const idsToDelete = new Set<string>();
    const idsToKeep = new Array();

    duplicates.forEach((group) => {
      
      const assetToKeep = suggestDuplicateByFileSize(group.assets);
      const assetsToDelete = group.assets.filter((asset) => asset.id !== assetToKeep.id);
      
      if (assetToKeep.exifInfo.dateTimeOriginal > '2017-01-01'
         && assetsToDelete.some((asset) => asset.originalFileName !== assetToKeep.originalFileName)
         && assetsToDelete.some((asset) => asset.exifInfo.dateTimeOriginal !== assetToKeep.exifInfo.dateTimeOriginal)) {
        return;
      }

      idsToKeep.push(assetToKeep);

      assetsToDelete.forEach((asset) => idsToDelete.add(asset.id));

      const albumIds = [...new Set(assetsToDelete.filter((asset) => assetIdToAlbums.has(asset.id)).flatMap((asset) => [...assetIdToAlbums.get(asset.id)]))];

      albumIds.forEach((albumId) => {
        if (assetIdToAlbums.has(assetToKeep.id) && assetIdToAlbums.get(assetToKeep.id)?.has(albumId)) {
          return;
        }

        albumIdToAddedAssets.get(albumId).add(assetToKeep.id);
      });

      if (!assetToKeep.isFavorite && assetsToDelete.some(asset => asset.isFavorite)) {
        newFavorites.push(assetToKeep.id);
      }
    });

    debugger;
    await updateAssets({assetBulkUpdateDto: { ids: newFavorites, isFavorite: true}});

    debugger;
    await Promise.all(albumIdToAddedAssets.entries().filter((entry) => entry[1].size > 0).map((entry) => addAssetsToAlbum(entry[0], [...entry[1]], false)));

    debugger;

    await deleteAssets({ assetBulkDeleteDto: { ids: Array.from(idsToDelete), force: !$featureFlags.trash } });
    debugger;
    await updateAssets({
      assetBulkUpdateDto: {
        ids: [...idsToDelete, ...idsToKeep.filter((id): id is string => !!id)],
        duplicateId: null,
      },
    });

    duplicates = [];

    deletedNotification(idsToDelete.length);

    return;
  };
```

mertalev replied:
Honestly, this should just be handled in the backend. All these queries and set operations can be handled *much* more efficiently with a single specialized DB query.

EinToni replied:
But if we automatically sync album membership while deduplicating single images, it should also be done when deduplicating all. Otherwise this would be quite unexpected behavior for the end user.

mertalev replied:
I'm not saying we shouldn't handle syncing when deduplicating all assets, just that the client is not a great place to be doing this. The DB is much better at doing these operations than JS, can do it without moving all of this data back and forth, and can make the entire process atomic. The bulk dedupe arguably should have been a server endpoint from the start, but with this extra layer of functionality it's just too much to keep in the client.

EinToni commented:
Hey,
after not doing anything for a month I just rebased onto main (214 commits!! crazy development).
Because I don't see this moving forward in a state like this I decided to do what @mertalev already suggested in the beginning. I split it up into the basic synchronization of favorite/archive/album and the more complex UI part. I will move the UI part into its own branch (unless someone contradicts), so this pull request here will only be for the basic automatic sync part.

I hope with this smaller scope and simpler functionality it will be possible to continue and maybe even merge this in the near future.

EinToni commented on i18n/en.json:
```diff
@@ -1237,6 +1238,12 @@
+  "synchronize_archives" : "Synchronize Archives",
+  "synchronize_archives_description" : "If one of the duplicates is a favorite, mark all deduplicated as favorites.",
+  "synchronize_favorites" : "Synchronize Favorites",
+  "synchronize_favorites_description" : "If one of the duplicates is already archived, also archive the deduplicated result.",
```
Fixed it

Phlogi commented:
Does this PR also sync metadata when mass-merging is happening through the action "Remove all duplicates"?

EinToni commented:
@Phlogi No it does not, currently this is only for the manual deduplication.

Phlogi commented:
> @Phlogi No it does not, currently this is only for the manual deduplication.

Thanks for clarifying, that would be a killer feature. The mass de-deplucation is almost useless without this. 

Is there a tagged docker image with your changes or would I need to build this to help testing? I don't think its built automatically.

NicholasFlamy commented:
> > @Phlogi No it does not, currently this is only for the manual deduplication.
> 
> Thanks for clarifying, that would be a killer feature. The mass de-deplucation is almost useless without this.
> 
> Is there a tagged docker image with your changes or would I need to build this to help testing? I don't think its built automatically.

Follow these instructions and for the clone the repo do `git clone https://github.com/EinToni/immich -b deduplicate-sync-album`
https://immich.app/docs/developer/setup#server-and-web-apps

NicholasFlamy commented:
> without the synchronization of the meta data, the mass deduplication is actually not usefully applicable.

I think there was an issue with mass dedupe recently, so that may be one of the obstacles before this can be approved.

EinToni commented:
@mertalev suggested to perform the metadata sync in the mass deduplication in the backend with DB queries. Unfortunately I am not proficient with databases, that's why I did not add it in. I would be happy if someone could add that.

rui-nar commented:
just wanted to drop a message saying that this feature is awesome and clearly mandatory. I have more than 2k duplicates with exactly this need.

@EinToni if you provide a fork of immich with an associated docker (or instructions on how to create one) i'll gladly use it :)

WolfgangDpunkt commented:
I can only agree with that.
Many users use Immich to merge their various photo collections. You upload photos from different sources and inevitably there are many duplicates.
However, if the metadata is destroyed or lost when cleaning up the duplicates, then this is not practical.

Therefore, I can only express my gratitude that this feature is being worked on.

EinToni commented:
Just rebased onto main, because there have been a lot new commits :D

@mertalev You suggested to do the mass deduplication on the server side and I fully agree, no need to hand all of the data back and forth for fully automated things. But I think this would be quite a stretch for this pull request. I would propose to open a separate issue for the backend mass dedupe with the sync of the metadata.

In the meantime the pull request got cut down to only merge the status of being archived/favorite and the album membership. Also here I would say extending this to more metadata could be done if future issues, or are single big pull requests preferred in immich?

If you don't mind, it would be great to also hear your thought about this @bo0tzz @alextran1502

alextran1502 commented:
Hello, thank you for rebasing and continuing the work here. I read the options and don't really have a clear understanding of what it does. So the copywriting will need some modification

Can you provide an example explanation for each option?

alextran1502 commented on web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:
```diff
@@ -188,6 +228,7 @@
         onclick={() => (isShowKeyboardShortcut = !isShowKeyboardShortcut)}
         aria-label={$t('show_keyboard_shortcuts')}
       />
+      <CircleIconButton icon={mdiCogOutline} title={$t('options')} onclick={() => (isShowOptions = !isShowOptions)} />
```
Please use the `<IconButton/>` component from the UI lib, similar to the button above this one

EinToni replied:
Replaced it 👍🏻

alextran1502 commented on web/src/lib/components/utilities-page/duplicates/duplicate-options.svelte:
```diff
@@ -0,0 +1,49 @@
+<FullScreenModal title={$t('options')} width="auto" {onClose}>
+  <div class="items-center justify-center">
+    <div class="grid p-2 gap-y-2">
+      <SettingSwitch
```
Please use this component instead https://ui.immich.app/components/switch

EinToni replied:
I replaced it for the switch and all over adjusted the FullScreenModal to fit into the design by adapting it equally to the recently reworked map-settings-modal.

EinToni commented:
> Hello, thank you for rebasing and continuing the work here. I read the options and don't really have a clear understanding of what it does. So the copywriting will need some modification
> 
> Can you provide an example explanation for each option?

Hey, thanks for the review. Unfortunately there was quite a bit of back and forth, therefore I get that it is confusing right now. Regarding the three options:
- Sync Albums: The selected deduplicated image will be put in all albums where the duplicated were.
  - If image `1` is in album `A`, image `2` in album `B` and image `3` is selected as the image to keep, it will be put in album `A` and `B`
- Sync Favorites: If any duplicate was marked as a favorite, the selected deduplicated image will be also marked as a favorite.
  - If image `1` is marked as favorite and image `2` is selected as the image to keep, it will be marked as a favorite.
- Sync Archive: If any duplicate was marked as archived, the selected deduplicated image will also be marked as archived.
  - If image `1` is marked as archived and image `2` is selected as the image to keep, it will be marked as archived.

Syncing more metadata would be great when deduplicating, but I guess that needs more discussion in terms of the implementation.

Spacey-Rat commented:
Looks great - would love to see this implemented. This missing feature to keep/copy missing album data keeps me away from de-dup'ing 15K images.

WolfgangDpunkt commented:
I wanted to emphasise the importance of continuing the great work on this issue. It's crucial for Immich's usefulness. :-)

Phlogi commented:
@EinToni  Can you rebase to main again, i'm happy to test this.

kacboy commented:
Would love to see this feature! I imported my Google Photos that have been compressed but would be very handy to replace existing albums with the original quality pics via this feature :)

Phlogi commented:
Could anyone rebase?

razor109 commented:
> Here's a rebase: https://github.com/NicholasFlamy/immich/tree/deduplicate-sync-album

I'm currently running a dockerized Immich 2.0.1 , how to best approach so i can use this rebase for cleaning up my library ? Can i just run a make on ubuntu and run this rebase parallel connected to the UPLOAD dir? any advice? 

Any insight if and when this would become part of main code?

NicholasFlamy commented:
> > Here's a rebase: https://github.com/NicholasFlamy/immich/tree/deduplicate-sync-album
> 
> I'm currently running a dockerized Immich 2.0.1 , how to best approach so i can use this rebase for cleaning up my library ? Can i just run a make on ubuntu and run this rebase parallel connected to the UPLOAD dir? any advice?
> 
> Any insight if and when this would become part of main code?

I wouldn't recommend blindly running it. I didn't have much time to test it or to fix it, but I think it might be slightly broken. Also, if you do want to try it, be sure you have a backup of your upload folder.

andreheuer commented:
@EinToni  or @NicholasFlamy 
Is there any chance to to rebase & merge the PR?
It would be reeeeeaaaallly great to have this feature. Using the current deduplication feature feels more destructive than helpful.

Phlogi commented:
Read the comments from the immich team. This PR won't be merged as it does not fit the architecture. The logic should be implemented in the backend, not frontend. 


https://github.com/immich-app/immich/pull/13851#discussion_r1933025477

EinToni committed:
Replace addAssetToAlbums with copyAsset

EinToni committed:
fix linter

EinToni commented on web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:
```diff
@@ -98,11 +108,68 @@
+          const idsToKeep = duplicateAssetIds.filter((id) => !trashIds.includes(id));
+          for (const sourceId of trashIds) {
+            for (const targetId of idsToKeep) {
+              await copyAsset({ assetCopyDto: { sourceId, targetId, albums: true } });
```
Oh that's nice. I thought the copy asset endpoint can only copy the albums, favorite status, complete sidecar file and stack  data from one asset to another. What I do here is synchronize the favorite status, visibility, rating, description and location between n-assets with a single `updateAssets` call.

When having the three assets:
- A, is marked as favorite
- B, has some location and description
- C, has some description and is the image to keep 

How would I go about synchronizing the data between these assets with the copy asset endpoint? Simply calling copy A -> C then B -> C would not work as then C would not be marked as a favorite as asset B is no favorite. I also assume the "sidecar" option of copy asset does exactly that, copy the sidecar file(?) The mentioned data does not have to be stored in the sidecar file but can also be embedded in the existing images.

It would be great if you could explain this further if I'm just missing something.

danieldietzler replied:
The copy asset endpoint only copies the sidecar file, i.e,. *metadata changes that have been made through Immich*. That is intentional, as we expect that original file metadata should always take precedence, unless you've made changes in Immich.

As for your described scenario, that's correct. I guess you'd need to figure out an order in which to merge them. 

Just in general however, if the logic gets too complex it should definitely not be in the +page.svelte file, and probably not even on web at all

Spacey-Rat commented:
Sorry... I followed this here and it somehow reads promising. But no idea what I need to do in order to use/check this. Or is it already implemented in the main release? I tried the duplicate checker recently but everything looks still the same. Sorry for not understanding everything in the commit process here...

bo0tzz commented:
@Spacey-Rat, this is a pull request, so a proposed code change. It's still open meaning it won't be available in anything on your end. If it's merged, it'll be available in the next release.

Phlogi commented:
I'm currently working on this one
- added merge of tags too
- extensive testing with 70+ test cases
- fixed album merging (we better use merge as a term than synchronizing, it's the common term)

planned/open:
- UI/UX improvements: if we merge the metadata, we don't need to show all of it of all candidates to the user. Decision should be on the image itself
- potentially move the service to the server backend, use API endpoint for de-duplication and metadata merge
- assess how risky the bulk de-duplication is with using CLIP score.

jrasm91 committed:
chore: clean up

jrasm91 reviewed (approved):
A few notes:
- I have found it to be problematic when implementation details are included in the translation strings themselves. If the implementation ever changes all the strings are now wrong and it's a bit of a pain to maintain. I think it's better to say "these things can be synchronized", but look at the docs for more information about _how_ exactly that works.
- I moved the modal to `src/modals`, and migrated it to the new `FormModal`
- I moved the preference to the new `PersistedLocalStorage` version
- It turns out we already have the full `AssetResponseDto` available from the original response, so I used that directly instead of re-fetching it again.

NicholasFlamy commented:
> 📖 Documentation deployed to [docs.pr-13851.preview.immich.app](https://docs.pr-13851.preview.immich.app)

Might want to make "Duplicate utility" title case in the docs: "Duplicate Utility"
